### DISCUSSION TITLE
A proposal for some additional fluent build methods

### DISCUSSION THREAD

**Discussion by StefH** [wiremock/WireMock.Net#1041] [Ideas]
Currently the C# code to build a request with a response is like:

``` csharp
server
    .Given(Request.Create()
        .WithPath("/mypath")
        .UsingPost()
    )
    
    .WithGuid("86984b0e-2516-4935-a2ef-b45bf4820d7d")
    
    .RespondWith(Response.Create()
        .WithHeader("Content-Type", "application/json")
        .WithBodyAsJson("{{JsonPath.SelectToken request.body \"..name\"}}")
        .WithTransformer()
    );
```

I want to make this a bit easier to use.

# :one: 
My first proposal is like this:
``` c#
server
    .WithRequest(r => r //⚡ The Request.Create() has been replaced by an Action
        .WithPath("/mypath")
        .UsingPost()
    )
    
    .DefineGuid("86984b0e-2516-4935-a2ef-b45bf4820d7d") //⚡ An extra method DefineGuid is added
    
    .RespondWith(r => r //⚡ The Response.Create() has been replaced by an Action
        .WithHeader("Content-Type", "application/json")
        .WithBodyAsJson("{{JsonPath.SelectToken request.body \"..name\"}}")
        .WithTransformer()
    );
```

# :two:
Optionally, my second proposal could be to use `WhenRequest`, so:
``` c#
server
    .WhenRequest(r => r //⚡ The Request.Create() has been replaced by an Action
        .WithPath("/mypath")
        .UsingPost()
    )
    
    .DefineGuid("86984b0e-2516-4935-a2ef-b45bf4820d7d") //⚡ An extra method DefineGuid is added
    
    .RespondWith(r => r //⚡ The Response.Create() has been replaced by an Action
        .WithHeader("Content-Type", "application/json")
        .WithBodyAsJson("{{JsonPath.SelectToken request.body \"..name\"}}")
        .WithTransformer()
    );
```

# :three:
Optionally, option 3, `RespondWith` is replaced by `Respond`:
``` c#
server
    .WhenRequest(r => r //⚡ The Request.Create() has been replaced by an Action
        .WithPath("/mypath")
        .UsingPost()
    )
    
    .DefineGuid("86984b0e-2516-4935-a2ef-b45bf4820d7d") //⚡ An extra method DefineGuid is added
    
    .Respond(r => r //⚡ The Response.Create() has been replaced by an Action
        .WithHeader("Content-Type", "application/json")
        .WithBodyAsJson("{{JsonPath.SelectToken request.body \"..name\"}}")
        .WithTransformer()
    );
```

Also another change would be that in case you just want a simple 200 ok response, you can now use:
``` c#
server
    .WithRequest(r => r
        .UsingGet()
    )
    .RespondWithOK();
```

or

``` c#
server
    .WithRequest(r => r
        .UsingGet()
    )
    .RespondWithStatusCode(300);
```

(https://github.com/WireMock-Net/WireMock.Net/pull/1042)

**Comment by StefH** [comment]
Tagging:
@cezarypiatek
@robertmircea
@MindaugasLaganeckas 
@MJLHThomassenHadrian
@mark-hingston
@basdijkstra 
@jflevesque-genetec
@sameena-ops 
@leolplex
@alastairtree
@NoahLerner
@phillee007
@vitaliydavydiak
@akamud
@kashifsoofi
@ggradnig
@eli-darkly
@cal-schleupen
@oleg-nenashev

**Reply by oleg-nenashev** [reply]
It would be definitely interesting! This is similar to what we do in Kotlin DSL and Golang

**Reply by sameena-ops** [reply]
I would like to have option like whenrequest(action,,)and respondwith(action,,)

**Reply by kashifsoofi** [reply]
My vote is for option 2, (a matter of opinion) for me `When...` and `ResponsdWith...` makes the intent clear. I also like the idea of Shortcut methods `RespondWithOK` and `ResponseWithStatusCode`.

**Reply by cezarypiatek** [reply]
I like the new API,  voting for option 2

**Reply by basdijkstra** [reply]
My vote goes to option 2 as well, with the suggestion to add alias methods `ThenRespondWith()` and similar to make it almost read like Gherkin.

Are you planning to create utility methods for all common status codes? Or just for `RespondWithOk()`? Because that would clutter the API in my opinion, I’d rather see you just offer `RespondWithStatusCode()` in that case.

**Reply by leolplex** [reply]
The option 2 reveals the intention, I'm inclined for second.

**Reply by jflevesque-genetec** [reply]
I follow @basdijkstra of having `ThenRespondWith` to follow the Given/When/Then format.

**Reply by StefH** [reply]
> My vote goes to option 2 as well, with the suggestion to add alias methods `ThenRespondWith()` and similar to make it almost read like Gherkin.
> 
> Are you planning to create utility methods for all common status codes? Or just for `RespondWithOk()`? Because that would clutter the API in my opinion, I’d rather see you just offer `RespondWithStatusCode()` in that case.

My intent is only to define `RespondWithOk()` as convenience method.

**Reply by cal-schleupen** [reply]
I also follow @basdijkstra of having ThenRespondWith to follow the Given/When/Then format. Thus, Option 2 ;-)

**Reply by MJLHThomassenHadrian** [reply]
I concur that the phrasing of option 2 conveys intent the best.

**Comment by jflevesque-genetec** [comment]
What is the `DefineGuid` used for?

**Reply by StefH** [reply]
> What is the `DefineGuid` used for?

It's the same as WithGuid.

This is used to assign a Guid to a mapping. If this is not defined WireMock will assign a random Guid.

**Comment by StefH** [comment]
Final version option :two:

``` c#
server
    .WhenRequest(r => r //⚡ The Request.Create() has been replaced by an Action
        .WithPath("/mypath")
        .UsingPost()
    )

    .DefineGuid("86984b0e-2516-4935-a2ef-b45bf4820d7d") //⚡ An extra method DefineGuid is added

    .ThenRespondWith(r => r //⚡ The Response.Create() has been replaced by an Action
        .WithHeader("Content-Type", "application/json")
        .WithBodyAsJson("{{JsonPath.SelectToken request.body \"..name\"}}")
        .WithTransformer()
        .WithStatusCode(HttpStatusCode.Accepted)
    );
```

**Comment by StefH** [comment]
PR https://github.com/WireMock-Net/WireMock.Net/pull/1042 has been merged. A new version 1.6.0 will be released in some days

**Comment by StefH** [comment]
I'll close this discussion for now because it's implemented.
